lib/private/Files/Cache/Cache.php

```diff
@@ -715,12 +715,29 @@ public function moveFromCache(ICache $sourceCache, $sourcePath, $targetPath) {
 					$query->set('encrypted', $query->createNamedParameter(0, IQueryBuilder::PARAM_INT));
 				}
 
-				try {
-					$query->execute();
-				} catch (\OC\DatabaseException $e) {
-					$this->connection->rollBack();
-					throw $e;
+				// Retry transaction in case of RetryableException like deadlocks.
+				// Retry up to 4 times because we should receive up to 4 concurrent requests from the frontend
+				foreach ([1, 2, 3, 4] as $i) {
+					try {
+						$this->connection->beginTransaction();
+						$query->execute();
+						break;
+					} catch (\OC\DatabaseException $e) {
+						$this->connection->rollBack();
+						throw $e;
+					} catch (RetryableException $e) {
+						// Simply throw if we already retried 4 times.
+						if ($i === 4) {
+							throw $e;
+						}
+
+						// Sleep for 1 seconde to give some time to the other transaction to finish.
+						sleep(1);
```

Comment:
I expect that there isn't a need to wait here, the deadlock error should clear both waiting locks I believe.

Reply:
Or I would at least reduce the sleep to 10-100ms

Reply:
I think it doesn't clear the original lock but a much shorter sleep should be enough.
We can also have an optimistic retry and wait longer after:
```suggestion
						usleep($i*$i*100);
```
I also think we should rollback before waiting